druntime/src/__importc_msvc_builtins.d

```diff
@@ -0,0 +1,12888 @@
+
+                    assert(((ushort(hi) << 8) | lo) == expectedResult);
+                    assert(unsignedMultiplyWithDoubleWidthProduct!(ubyte, true)(left, right) == hi);
+
+                    ++right;
+                }
+                while (right != 0);
+
+                ++left;
+            }
+            while (left != 0);
+        }
+
+        /* This is trusted so that it's @safe without DIP1000 enabled. */
+        pragma(inline, true)
+        private I signedMultiplyWithDoubleWidthProduct(I, bool onlyHighHalf)(
+            I low,
+            I high,
+            scope AliasSeq!(I*)[0 .. !onlyHighHalf] highProduct
+        ) @trusted pure nothrow @nogc
+        if (__traits(isIntegral, I) && !__traits(isUnsigned, I))
+        {
+            import core.bitop : bsr;
+
+            alias UnsignedI = AliasSeq!(ubyte, ushort, uint, ulong)[I.sizeof.bsr];
+
+            UnsignedI lo = cast(UnsignedI) low;
+            UnsignedI hi = cast(UnsignedI) high;
+
+            static if (onlyHighHalf)
+            {
+                hi = unsignedMultiplyWithDoubleWidthProduct!(UnsignedI, true)(lo, hi);
+            }
+            else
+            {
+                lo = unsignedMultiplyWithDoubleWidthProduct!(UnsignedI, false)(lo, hi, &hi);
+            }
+
+            hi -= high * (low < 0);
+            hi -= low * (high < 0);
+
+            static if (onlyHighHalf)
+            {
+                return hi;
+            }
+            else
+            {
+                *highProduct[0] = hi;
+                return lo;
+            }
+        }
+
+        @safe pure nothrow @nogc unittest
+        {
+            /* The mechanics used to get a double-width product from two operands are the same regardless of the width.
+               So, if this works for 8x8->16-bit multiplication, it'll work for 64x64->128-bit multiplication. */
+
+            byte left = byte.min;
+            byte right = byte.min;
+
+            do
+            {
+                do
+                {
+                    short expectedResult = left * right;
+
+                    byte hi = left;
+                    byte lo = right;
+                    lo = signedMultiplyWithDoubleWidthProduct!(byte, false)(lo, hi, &hi);
+
+                    assert(cast(short) (((short(hi) << 8) & 0xFF00) | (lo & 0x00FF)) == expectedResult);
+                    assert(signedMultiplyWithDoubleWidthProduct!(byte, true)(left, right) == hi);
+
+                    ++right;
+                }
+                while (right != byte.min);
+
+                ++left;
+            }
+            while (left != byte.min);
+        }
+    }
+
+    version (X86_64)
+    {
+        extern(C)
+        pragma(inline, true)
+        long _div128(long highDividend, long lowDividend, long divisor, scope long* remainder) @safe pure nothrow @nogc
+        {
+            if (__ctfe)
+            {
+                /* This is an amalgamation of core.int128.divmod and core.int128.neg.  */
```

Comment:
why not simply forward to that implementation?

Reply:
I think that would rely on linking with DRuntime, which would rule out BetterC compatibility.